Prevent call to Item#initializeClient when in datagen

sciwhiz12 opened the pull request:
When running under data-generation, the `Minecraft` singleton is unavailable even though the operating env is under the `CLIENT` dist. Along with the fact that most uses of IItemRenderProperties (usually to provide a custom BEWLR) make use of the `Minecraft` singleton, this means that developers unaware of `Minecraft`'s unavailability will encounter a null pointer exception when they try to access its singleton.

This PR adds a check to prevent the `Item#initializeClient` method from being called during data-generation, to prevent the situation described before from causing an NPE. As `IItemRenderProperties` has nothing which might be used by datageneration, this is safe to do.

This PR is inspired by a [Discord conversation][discord] on The Forge Project discord server wherein @gigaherz expressed that this method shouldn't be called in datagen but this check was forgotten, and that it should be PR-ed.

[discord]: https://discord.com/channels/313125603924639766/867851603468615740/888932184122069036

sciwhiz12 committed:
Prevent call to Item#initializeClient when in datagen

When running under data-generation, the Minecraft singleton is unavailable
even though the operating env is under the CLIENT dist. Adding this check
prevents crashes from devs unaware of this, as most uses of
IItemRenderProperties (usually for a custom BEWLR) make use of the
Minecraft singleton and therefore will cause NPEs if not guarded against.

gigaherz reviewed (changes requested):
There's 3 initializeClient methods in total (Item, Block, and MobEffect). Would be nice to get the same fix done to the other 2.

marchermans commented:
> initializeClient

I noticed them, but I am not sure they are related to issues in Datagen.
How are these other two involved?

gigaherz commented:
> How are these other two involved?

They are all equally involved. When datagen runs, forge runs its initialization logic, during which it registers all the things (items, blocks, and mobeffects included).
Since the initializeClient calls are done in the constructor of their respective object, it means all 3 will be called during datagen, and all 3 will run into the same issue if the logic in those classes references Minecraft.getInstance().

sciwhiz12 committed:
Prevent more calls to initializeClient during datagen

This patches MobEffect and Block with the same patch for
Item#initializeClient.

sciwhiz12 commented:
Updated to also prevent `Block#initializeClient` and `Item#initializeClient` from being called during datagen.

gigaherz commented:
Oops I forgot about this. Meant to merge sometime I wasn't busy.... and I guess I'm not busy now.
